mut/index/SnootyManifest.py

```diff
@@ -0,0 +1,215 @@
+        # with Product (DOP-)
+        logger("\tDetermining indexability")
+        noindex = False
+        reasons: List[str] = []
+        # If :robots: None in metadata, do not index
+        if self.robots == False:
+            noindex = True
+            reasons.append("robots=None in meta directive")
+
+        # If page has no title, do not index.
+        if self.title == None:
+            noindex = True
+            reasons.append("This page has 0 headings, not even the H1")
+
+        logger("\tnoindex: {}".format(noindex))
+        return noindex, reasons
+
+    def export(self) -> Optional[Dict[str, Any]]:
+        '''Generate the manifest dictionary entry from the AST source.'''
+
+        if self.noindex:
+            logger("\tRefusing to index {} because: {}".format(self.slug, ' '.join(self.reasons)))
+            return
+
+        document = {
+            "slug": self.slug,
+            "title": self.title,
+            "headings": self.headings,
+            "paragraphs": self.paragraphs,
+            "code": self.code,
+            "preview": self.preview,
+            "tags": self.keywords,
+        }
+
+        return document
+
+class Manifest:
+    '''Manifest to provide to Atlas search.'''
+    def __init__(self, url: str, includeInGlobalSearch: bool) -> None:
+        self.url = url
+        self.globally = includeInGlobalSearch
+        self.documents: List[Dict[str, Any]] = [] 
+
+    def add_document(self, document: Dict[str, Any]) -> None:
+        '''Add a document to the manifest'''
+        if not document == None:
+            self.documents.append(document)
+    
+    def export(self) -> str:
+        '''Return the manifest as json.'''
+        manifest = {
+            "url": self.url,
+            "includeInGlobalSearch": self.globally,
+            "documents": self.documents
+        }
+        return dumps(manifest, indent=4)
+
+def process_snooty_manifest_bson(path: str) -> Any:
+    '''Generates manifest info for a BSON document.'''
+    with open(path, 'rb') as f:
+        logger(path)
+        data = decode_all(f.read())
+        document = Document(data).export()
+    return document
+
+def generate_manifest(ast_source: str, url: str, includeInGlobalSearch: bool) -> str:
+    '''Process BSON files and compile a manifest.'''
+    manifest = Manifest(url, includeInGlobalSearch)
+    print("Building manifests for {} documents".format(len(ast_source)))
+
+    with ProcessPoolExecutor() as executor:
+        for bson_doc in executor.map(process_snooty_manifest_bson, ast_source):
+            manifest.add_document(bson_doc)
+        return manifest
+
+def get_ast_list(walk_dir: str) -> List[str]:
+    '''Get full list of BSON paths that need to be processed,
+       get rid of files that don't need to be indexed, but exist as AST.'''
+    ast_source_paths: List[str] = []
+
+    for root, dirs, files in walk(walk_dir):
+        if 'images' in dirs:
```

Comment:
I'd prefer to make this less branch-y. e.g.
```
for forbidden_name in ["images", "includes", "sharedinclude", ".DS_Store"]:
    if forbidden_name in dirs:
        dirs.remove(forbidden_name)
```

If you want to get FANCY, you could even do a set:
```
for forbidden_name in set(["images", "includes", "sharedinclude", ".DS_Store").intersection(dirs):
    dirs.remove(forbidden_name)
```

Reply:
LOVE IT.

Reply:
SO MUCH CLEANER